
# wrnchAI Hand's on

## Objectives

At the end of this presentation you should be able to install and use wrnchAI

- What's wrnch?
    - Live example
    - Introduction to pose estimation
    - Joints definition

- Getting started with:
    - Cloud REST API
    - Edge SDK

This getting started notebook will be shared to all participants

# Whats wrnch?

## Human pose estimation
- 2d, 3d, head and hand pose estimations
- Activity recognition (with Edge SDK)

## How to use wrnch
- Cloud: REST API
- Edge: `Python` & `C++` API 

---
# Getting started with joint definition 

## 1: Register and test wrnchAI
### Register you developer account

- Register throught **this link to get a 5 days free dev license** : https://devportal.wrnch.ai/codejam2019
- NB: if you don't use this URL, you **won't** receive a free license
    



### Submit an image or a video
- For this demo we will use this image:


![test-image](img/small/test-image.jpg)


- Once register you can submit video to our test page to this address:
    https://devportal.wrnch.ai/wrnchcloud

- Ask for both JSON and Annotated media with those options:
    - Head Pose Estimation
    - Hand Pose Estimation

# 2: Retrieve results

Once our servers have processed the image, download the results. If you have selected more than one worktype, you will receive them as a zip. In our example the zip has an annotated_media and a json file

The annotated media will look like this:

![result img](img/small/test-image-processed.jpeg)


# 3. Process the results

## Our goal
-> Lets verify that his left wrist is on his left knee

### 1. Import the json file
 

In [3]:
import json

def json_r(filename):
   with open(filename) as f_in:
       return(json.load(f_in))

pose_estimation = json_r('test-image-processed.json')

print(str(pose_estimation['frames'])[:300])

[{'frame_time': 0, 'height': 819, 'persons': [{'hand_pose': {'left': {'bbox': {'height': 0.2857142984867096, 'minX': 0.6681647300720215, 'minY': 0.3588320314884186, 'width': 0.29249998927116394}, 'is_main': False, 'joints': [0.7872003316879272, 0.49528399109840393, 0.7665179371833801, 0.510883808135


In [4]:
man_pose = pose_estimation['frames'][0]['persons'][0]

print(man_pose.keys())

dict_keys(['hand_pose', 'head_pose', 'id', 'pose2d'])


In [5]:
import pprint
# Lets print the 10 first joint coordinates
pprint.pprint(man_pose['pose2d']['joints'][0:10])

[0.3711889684200287,
 0.7936080694198608,
 0.5704107880592346,
 0.7095877528190613,
 0.3788127601146698,
 0.6029024720191956,
 0.5351067781448364,
 0.5762162208557129,
 0.7694329023361206,
 0.5570101141929626]


### 2. Interpret the coordinates
 
 Each joint has a 2d coordinate in this space: 
 
 ![joint coor](img/small/with-arrow.png)
 
 #### We define both 2d and 3d coordinate on this page: https://devportal.wrnch.ai/wrnchai_sdk/coord_spaces

### Joints definition
#### For more information and figure, refer to our joint definition documentation: 
https://devportal.wrnch.ai/wrnchai_sdk/joint_definitions


![hands joint doc](img/small/j25_body_joints.jpg)


#### There are 25 joints. The json should have 50 coordinates

In [6]:
print(len(man_pose['pose2d']['joints']))

50


Lets determine if his left wrist is over his left knee.
We should have :

`xLWRIST ~= xLKNEE`

In [7]:
xLWRIST = man_pose['pose2d']['joints'][30]
xLKNEE = man_pose['pose2d']['joints'][8]



print(xLWRIST)
print(xLKNEE)

0.7735595107078552
0.7694329023361206


In [13]:
# You can call this in the main just to get the array which has X and Y co-ordinate of the joint
def getJoint(jointX, jointY):
    return [man_pose['pose2d']['joints'][jointX],man_pose['pose2d']['joints'][jointY]]


RANKLE = getJoint(0,1)

0.3711889684200287
0.5704107880592346


In [69]:
# pass man_pose['pose2d']['joints']
# returns an array minXVal, minYVal, maxXVal, maxYVal
def getMinMax(joints):
    minYVal = 10000000
    minXVal = 10000000
    maxXVal = 0
    maxYVal = 0
    for i in range(0, len(joints), 2):
        if joints[i] > 0:
            if joints[i] < minXVal:
                minXVal = joints[i]
            if joints[i] > maxXVal:
                maxXVal = joints[i]

            
    for i in range(1, len(joints), 2):
        if joints[i] > 0:
            if joints[i] < minYVal:
                minYVal = joints[i]
            if joints[i] > maxYVal:
                maxYVal = joints[i]
                   
    return [minXVal, minYVal, maxXVal, maxYVal]
        
    
    

In [70]:
def normalize(joints, croppedImage,baseWidth, baseHeight):

    normalizedImage = [None] * len(joints)

    for i in range(0, len(joints), 2):
        normalizedImage[i] = joints[i] - croppedImage[0]
        
    for i in range(1, len(joints), 2):
        normalizedImage[i] = joints[i] - croppedImage[1]
    
    currentWidth = croppedImage[2] - croppedImage[0]
    currentHeight = croppedImage[3] - croppedImage[1]

    for i in range(0, len(normalizedImage), 2):
        normalizedImage[i] = normalizedImage[i]*(baseWidth/currentWidth)
    
    for i in range(1, len(normalizedImage), 2):
        normalizedImage[i] = normalizedImage[i]*(baseHeight/currentHeight)
        
    return normalizedImage
    

In [126]:
# import math 
# def partitionImage(normalizedImage, cellNums):
#     minmax = getMinMax(normalizedImage)
#     width = minmax[2]-minmax[0]
#     print("width is: dsjdnjknfdfd" , width)
#     height = minmax[3]-minmax[1]
#     print("heightttttttttttttttttt", height)
#     area = width * height
    
#     partitionedArray = [0] * cellNums
#     cellArea = area / cellNums
#     cellLength = math.sqrt(cellArea)
    
#     startLengthX = 0
#     endLengthX = cellLength
#     startLengthY = 0
#     endLengthY = cellLength
    
#     cellsPerRow = width/cellLength
    
#     for i in range(0, len(normalizedImage), 2):
#         print ("Normalized point:")
#         print (normalizedImage[i])
#         print (";")
#         print (normalizedImage[i+1])

#         for j in range(0,cellNums):
#             print ("cell")
#             print ("startLengthX")
#             print (startLengthX)
#             print ("startLengthY")
#             print (startLengthY)
            
#             print ("endLengthX")
#             print (endLengthX)
#             print ("endLengthY")
#             print (endLengthY)
#             print("************")
            
#             if(startLengthX <normalizedImage[i] < endLengthX):
#                 if(startLengthY <normalizedImage[i+1] < endLengthY):
#                     print ("FOUND CELL!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
#                     partitionedArray[j] = 1
#                     break;
                        
#             startLengthX = endLengthX
#             endLengthX+=cellLength
#             print ("left ineequality")
#             print((j % cellsPerRow) * cellLength )
#             if(startLengthX >= width):
#                 startLengthX = 0
#                 endLengthX=cellLength
#                 startLengthY = endLengthY
#                 endLengthY+=cellLength
                
#             if (startLengthY >= height)
#                 startLengthY = 0
#                 endLengthY = 0

                
    
#     return partitionedArray
                
    

In [172]:
import math
def partitionImage(normalizedImage, cellNums):
    minmax = getMinMax(normalizedImage)
    width = minmax[2]-minmax[0]
    height = minmax[3]-minmax[1]
    area = width * height
    
    partitionedArray = [0] * cellNums
    cellArea = area / cellNums
#     cellArea = 0.00001
    cellLength = math.sqrt(cellArea)
    /
    upperLeftX = 0
    upperLeftY = 0
    lowerLeftX = 0
    lowerLeftY = cellLength
    upperRightX = cellLength
    upperRightY = 0
    lowerRightX = cellLength
    lowerRightY = cellLength
    
    print(cellLength)
    
    for i in range(0, len(normalizedImage), 2):
        for j in range(0,cellNums):
            print("upperleftX",upperLeftX)
            print("upperleftY",upperLeftY)
            print("lowerleftX",lowerLeftX)
            print("lowerleftY",lowerLeftY)
            print("upperRightX",upperRightX)
            print("upperRightY",upperRightY)
            print("lowerRightX",lowerRightX)
            print("lowerRightY",lowerRightY)
            print(normalizedImage[i])
            print(normalizedImage[i+1])
            print("*****************************")
            if(check(upperRightX,upperRightY, lowerRightX,lowerRightY,lowerLeftX, lowerLeftY, upperLeftX, upperLeftY,normalizedImage[i],normalizedImage[i+1])):
                partitionedArray[i]=1
#                 upperLeftX = 0
#                 upperLeftY = 0
#                 lowerLeftX = 0
#                 lowerLeftY = cellLength
#                 upperRightX = cellLength
#                 upperRightY = 0
#                 lowerRightX = cellLength
#                 lowerRightY = cellLength
                
                
            upperLeftX = upperRightX
            lowerLeftX = lowerRightX
            upperRightX += cellLength
            lowerRightX += cellLength
            
            if(upperRightX >= width):
                upperLeftX = 0
                lowerLeftX = 0
                upperRightX = cellLength
                lowerRightX = cellLength
                upperLeftY+=cellLength
                upperRightY+=cellLength
                lowerRightY += cellLength
                lowerLeftY += cellLength
            
            
    
    return partitionedArray

In [173]:
def area(x1, y1, x2, y2, x3, y3): 
      
    return abs((x1 * (y2 - y3) + 
                x2 * (y3 - y1) + 
                x3 * (y1 - y2)) / 2.0) 

def check(x1, y1, x2, y2, x3,  
          y3, x4, y4, x, y): 
                
    # Calculate area of rectangle ABCD  
    A = (area(x1, y1, x2, y2, x3, y3) +
         area(x1, y1, x4, y4, x3, y3)) 
  
    # Calculate area of triangle PAB  
    A1 = area(x, y, x1, y1, x2, y2) 
  
    # Calculate area of triangle PBC  
    A2 = area(x, y, x2, y2, x3, y3) 
  
    # Calculate area of triangle PCD  
    A3 = area(x, y, x3, y3, x4, y4) 
  
    # Calculate area of triangle PAD  
    A4 = area(x, y, x1, y1, x4, y4); 
  
    # Check if sum of A1, A2, A3  
    # and A4 is same as A  
    return (A == A1 + A2 + A3 + A4)

In [174]:
croppedImage = getMinMax(man_pose['pose2d']['joints'])
normalizedImage = normalize(man_pose['pose2d']['joints'], croppedImage, 100,200)
# print (normalizedImage)
partitionImages = partitionImage(normalizedImage, 10)
print(partitionImages)


43.182138580828536
upperleftX 0
upperleftY 0
lowerleftX 0
lowerleftY 43.182138580828536
upperRightX 43.182138580828536
upperRightY 0
lowerRightX 43.182138580828536
lowerRightY 43.182138580828536
7.593297715625786
162.2284914048556
*****************************
upperleftX 43.182138580828536
upperleftY 0
lowerleftX 43.182138580828536
lowerleftY 43.182138580828536
upperRightX 86.36427716165707
upperRightY 0
lowerRightX 86.36427716165707
lowerRightY 43.182138580828536
7.593297715625786
162.2284914048556
*****************************
upperleftX 0
upperleftY 43.182138580828536
lowerleftX 0
lowerleftY 86.36427716165707
upperRightX 43.182138580828536
upperRightY 43.182138580828536
lowerRightX 43.182138580828536
lowerRightY 86.36427716165707
7.593297715625786
162.2284914048556
*****************************
upperleftX 43.182138580828536
upperleftY 43.182138580828536
lowerleftX 43.182138580828536
lowerleftY 86.36427716165707
upperRightX 86.36427716165707
upperRightY 43.182138580828536
lowerRightX

lowerleftY 1554.5569889098267
upperRightX 86.36427716165707
upperRightY 1511.3748503289983
lowerRightX 86.36427716165707
lowerRightY 1554.5569889098267
39.918319716212466
42.88113001944973
*****************************
upperleftX 0
upperleftY 1554.5569889098267
lowerleftX 0
lowerleftY 1597.7391274906552
upperRightX 43.182138580828536
upperRightY 1554.5569889098267
lowerRightX 43.182138580828536
lowerRightY 1597.7391274906552
39.918319716212466
42.88113001944973
*****************************
upperleftX 43.182138580828536
upperleftY 1554.5569889098267
lowerleftX 43.182138580828536
lowerleftY 1597.7391274906552
upperRightX 86.36427716165707
upperRightY 1554.5569889098267
lowerRightX 86.36427716165707
lowerRightY 1597.7391274906552
39.918319716212466
42.88113001944973
*****************************
upperleftX 0
upperleftY 1597.7391274906552
lowerleftX 0
lowerleftY 1640.9212660714836
upperRightX 43.182138580828536
upperRightY 1597.7391274906552
lowerRightX 43.182138580828536
lowerRightY 1640

upperRightX 86.36427716165707
upperRightY 3281.842532142965
lowerRightX 86.36427716165707
lowerRightY 3325.0246707237934
94.16141457761277
88.52305458076533
*****************************
upperleftX 0
upperleftY 3325.0246707237934
lowerleftX 0
lowerleftY 3368.206809304622
upperRightX 43.182138580828536
upperRightY 3325.0246707237934
lowerRightX 43.182138580828536
lowerRightY 3368.206809304622
94.16141457761277
88.52305458076533
*****************************
upperleftX 43.182138580828536
upperleftY 3325.0246707237934
lowerleftX 43.182138580828536
lowerleftY 3368.206809304622
upperRightX 86.36427716165707
upperRightY 3325.0246707237934
lowerRightX 86.36427716165707
lowerRightY 3368.206809304622
94.16141457761277
88.52305458076533
*****************************
upperleftX 0
upperleftY 3368.206809304622
lowerleftX 0
lowerleftY 3411.3889478854503
upperRightX 43.182138580828536
upperRightY 3368.206809304622
lowerRightX 43.182138580828536
lowerRightY 3411.3889478854503
94.16141457761277
88.5230

upperleftY 5181.856629699429
lowerleftX 43.182138580828536
lowerleftY 5225.0387682802575
upperRightX 86.36427716165707
upperRightY 5181.856629699429
lowerRightX 86.36427716165707
lowerRightY 5225.0387682802575
63.019407002037354
188.0337660422826
*****************************
upperleftX 0
upperleftY 5225.0387682802575
lowerleftX 0
lowerleftY 5268.220906861086
upperRightX 43.182138580828536
upperRightY 5225.0387682802575
lowerRightX 43.182138580828536
lowerRightY 5268.220906861086
63.019407002037354
188.0337660422826
*****************************
upperleftX 43.182138580828536
upperleftY 5225.0387682802575
lowerleftX 43.182138580828536
lowerleftY 5268.220906861086
upperRightX 86.36427716165707
upperRightY 5225.0387682802575
lowerRightX 86.36427716165707
lowerRightY 5268.220906861086
63.019407002037354
188.0337660422826
*****************************
upperleftX 0
upperleftY 5268.220906861086
lowerleftX 0
lowerleftY 5311.403045441915
upperRightX 43.182138580828536
upperRightY 5268.220906861

In [164]:
# This is the difference between right knee and right ankle
RdiffAnkleKnee = [RANKLE[0] - RKNEE[0],RANKLE[1] - RKNEE[1]]
# This is the difference between right hip and right knee
RdiffKneeHip = [RKNEE[0] - RHIP[0],RKNEE[1] - RHIP[1]]

# This is the difference between left knee and left ankle
LdiffAnkleKnee = [LANKLE[0] - LKNEE[0],LANKLE[1] - LKNEE[1]]
# This is the difference between left hip and left knee
LdiffKneeHip = [LKNEE[0] - LHIP[0],LKNEE[1] - LHIP[1]]


-> Indeed the wrist and the elbow are close in the frame

RAnkle = [man_pose['pose2d']['joints'][0],man_pose['pose2d']['joints'][1]]

print(RAnkle[0])


# Getting started with the cloud API
All of this is also documented here: https://devportal.wrnch.ai/wrnchcloud/api_docs#introduction

First install requests. The python http client.

In [6]:
!pip install requests
import requests
import json

You are using pip version 19.0.3, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


The two paths we will need for this example

In [7]:
LOGIN_URL = 'https://api.wrnch.ai/v1/login'
JOBS_URL = 'https://api.wrnch.ai/v1/jobs'

1. Retrieve the cloud API key

Click on the key icon on this page: https://devportal.wrnch.ai/licenses to find your cloud API key

In [15]:
#Save the Cloud API key for next cell to work
API_KEY = ""

2. Authenticate

In [9]:
resp_auth = requests.post(LOGIN_URL,data={'api_key':API_KEY})
print(resp_auth.text)
# the jwt token is valid for an hour
JWT_TOKEN = json.loads(resp_auth.text)['access_token']

{"access_token": "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOjc1NiwibGljZW5zZV9pZCI6NTkxLCJhZGRvbnMiOnt9LCJleHAiOjE1NzM4MzkyNjMsImlkZW50aXR5Ijo3NTYsInJvbGUiOiJpbmRfY2xpZW50IiwiaWF0IjoxNTczODMyMDYzLCJqdGkiOiJhNzYyYzBiYWVmZmVmMmUzY2Y1ZTllMmM5Y2YxNjdhNDE5M2I5YzcxMDBiZTUyNDhmMzcyZDBiNzQxNWVjNDMwIiwidHlwZSI6ImFjY2VzcyIsImZyZXNoIjoiZmFmIn0.OOWitopW0bsZ0Ne5hLsyhh3PzctKvBm_HLwUeYYXLQ4"}



Now lets submit a file for processing
1. Open the file as a byte stream
2. Send a post request with authentification and the file

In [10]:
with open('img/test-image.jpg', 'rb') as f:
    resp_sub_job = requests.post(JOBS_URL,
                                 headers={'Authorization':f'Bearer {JWT_TOKEN}'},
                                 files={'media':f},
                                 data={'work_type':'json'}
                                )

In [11]:
job_id = json.loads(resp_sub_job.text)['job_id']
print('Status code:',resp_sub_job.status_code)
print('Response:',resp_sub_job.text)

Status code: 202
Response: {"job_id": "1cc5c026-1ac8-483b-8ff5-83ec09fa425e"}



The status code should be 202 and return a job_id.

Now lets retrieve the result

In [12]:
GET_JOB_URL = JOBS_URL + '/' +job_id
print(GET_JOB_URL)
resp_get_job = requests.get(GET_JOB_URL,headers={'Authorization':f'Bearer {JWT_TOKEN}'})
print('Status code:',resp_get_job.status_code)
print('\nResponse:',resp_get_job.text)

https://api.wrnch.ai/v1/jobs/1cc5c026-1ac8-483b-8ff5-83ec09fa425e
Status code: 200

Response: {"file_info":{"joint_definitions":{"hands":"hand21","head":"wrFace20","pose2d":"j25","pose3d_ik":"extended","pose3d_raw":"j25"}},"frames":[{"frame_time":0,"height":614,"persons":[{"id":0,"pose2d":{"bbox":{"height":0.9895586371421814,"minX":0.2897970974445343,"minY":0.039630815386772156,"width":0.5532262325286865},"is_main":true,"joints":[0.37112537026405334,0.7902586460113525,0.5703703165054321,0.7099169492721558,0.3749939203262329,0.6030898690223694,0.5311734676361084,0.5763961672782898,0.7657291889190674,0.5610739588737488,0.6757756471633911,0.866474986076355,0.45308369398117065,0.5897430181503296,0.5196003913879395,0.2996460795402527,0.5196273922920227,0.29780837893486023,0.5455899834632874,0.13356269896030426,0.5350296497344971,0.5954576730728149,0.3633612096309662,0.46940380334854126,0.42190080881118774,0.28249606490135193,0.6172999143600464,0.31679609417915344,0.6875200271606445,0.458121

Lets see if the left wrist and left knee are touching:

In [13]:
cloud_pose_estimation = json.loads(resp_get_job.text)
cloud_xLWRIST = cloud_pose_estimation['frames'][0]['persons'][0]['pose2d']['joints'][30]
cloud_xLKNEE = cloud_pose_estimation['frames'][0]['persons'][0]['pose2d']['joints'][8]

print(cloud_xLWRIST)
print(cloud_xLKNEE)

0.7735621333122253
0.7657291889190674


-> Now you can use human pose estimation in your project with very few line of code

---

# Getting started with the edge sdk

The Edge SDK allows you to do real-time pose estimation.

For installation procedure refer to the detailed documentation here: 
https://devportal.wrnch.ai/wrnchai_sdk/releases

NB: We do not support MacOS yet.

Common pitfalls when working with the Python SDK
- `No module called wrnchAI` : Ensure that lib/python/X.X is in your PYTHONPATH. Or simply, on the top of your python script do
  ```python
  import sys
  sys.path.append('<path_to_project_root>/lib/python/X.X')
  ```
- License problems : Ensure the license is under your home directory or set as an environment variable
- When working with python, it's good practice to create a virtual environment.

For example with Python3:

   - `python -m venv <env_name>`

   - Then activate it by running the activate script in the bin directory
    
   - Then add opencv dependency for the samples with this command: `pip install opencv-python`
